### DISCUSSION TITLE
Functionality "lost" in SAPC/APCA

### DISCUSSION THREAD

**Discussion by front-endian** [Myndex/SAPC-APCA#57] [Ideas]
Howdy!

First of all, thank you for all of the work you've put into perceptual contrast measurements. We've started incorporating APCA measurements into some of our design work and have found the beta versions to be very helpful for informing design decisions. In that work we found one use case where SAPC/APCA lost a certain degree of functionality WCAG's measurement system had:

It is common to use color to show different states, whether it is button types, graphs, banner messages, etc. In these cases, the color contrast you are comparing are not foreground/background and you aren't measuring readability of text. The "contrast" you care about in these cases is how easily can a user see that one element is a different color than another. Obviously all these designs need text, icons, or other non-color elements to help users distinguish them, but good color contrast is important to ensure all users can distinguish the elements at a glance without regardless of what their color vision is like.

While generally inaccurate for representing human perception, WCAG's symmetry for foregrounds/backgrounds made it useful for these contrast measurements when it was all we had.

- When using SAPC/APCA in their current forms for this use case, should we swap the foreground/background values and take the average? Use the lower of the two values?
- Is this a use case SAPC/APCA can support? Something research could be done around?

Thank you again for all your work!

**Comment by TaigaYamada** [comment]
#38 May be helpful

**Reply by front-endian** [reply]
Thank you!

**Comment by Myndex** [comment]
Hi Colin! @justcolin 

Thank you for two great questions, sorry it's taken a bit for me to respond, for some reason GitHub did not notify me.

> _It is common to use color to show different states, whether it is button types, graphs, banner messages, etc. In these cases, the color contrast you are comparing are not foreground/background and you aren't measuring readability of text._ 

Yes, so the input that its labeled "text" is the "stimuli" field, and the one labeled "background" is for the surrounding color area. They are only labeled text and background to keep the learning curve short. The full SAPC and SARCAM appearance models deal with much more than just a pair of colors. Here it is kept simple for guideline purposes.

Basically, the "thing" you want people to _notice_ goes in the "text" field and the stuff that surrounds it goes in the background. I discuss use cases and categories of use [in this discussion post.](https://github.com/Myndex/SAPC-APCA/discussions/39#discussion-3757102)

Take a look at that and we can certainly continue with questions in this thread.

> _The "contrast" you care about in these cases is how easily can a user see that one element is a different color than another. Obviously all these designs need text, icons, or other non-color elements to help users distinguish them..._

Spatial frequency is the dominant factor for small/thin elements, while the contrasts of large thick elements and areas of color are dominated by luminance contrast, which in the case of APCA is calculated as the difference of perceptual lightness/darkness.

### Spatial Case
While it's clear that the high spatial frequencies of fonts make them a special case for readability, non-text is also subject to spatial frequency effects. A principal difference is that you'll be much more likely to have very low spatial frequencies for non text elements.

Consider a button, not including the text, or form fields — low spatial frequencies by which I mean larger than 4px in stroke width OR for a filled element the smallest dimension is larger than ~4px. These large elements do not need the same level of luminance contrast.

Here's a preliminary table of Lc values to non-text element size:

   <img width="300" alt="APCA non text minimums Feb 2 2022 — this is a table of Lc values to elements size in px" src="https://user-images.githubusercontent.com/42009457/152306316-18f7dcef-244a-44c9-a2dd-3fa07b9c698f.png">

I'm going to look at your other message, but do feel free to continue in this thread with questions.


Thank you!

Andy

**Reply by front-endian** [reply]
Thank you for the lengthy response. This is all helpful information, but I'm not sure it answers the original question.

As a common example, consider a very simple website with only plain text and links. There are three places where contrast matters:

- Contrast of the plain text against the background
- Contrast of the links against the background
- Contrast between the plain text and the links

The first two are what APCA seems to cover well: the contrast of objects against a background needed for someone to clearly read text or distinguish what an icon is.

The third APCA doesn't seem to let us measure as well. Neither of the two colors is the thing you want to notice, nor does one surround the other. Both colors are of equally important and on seperate parts of the display.

To satisfy WCAG 2 Use of Color (1.4.1) we need to use something other than color to distinguish the links from the plain text. This could be an underline, but it is also considered passing if there is a 3:1 contrast ratio between the two colors. If you follow the preferred method and add a completely non-color indicator like an underline, a good design would still have a nice color contrast between the links and the plain text to make links easier to distinguish/find.

While we can try to pick good colors by eye, having an empirical way to measure this type of contrast would be extremely helpful.

To rephrase the original questions:
- Is it possible to user APCA or SAPC in their current state to perform these kinds of contrast measurements?
- In the example case of using color to distinguish links from body text what is the equivalent of the old 3:1 minimum contrast recommendation to ensure that users can see that they are two different colors?
  - Lc 15 because that is the minimum for something to be considered discernible?
  - Lc 30?
  - Lc 45 because that is considered the minimum for icons with finer details?
- If a non-color indicator (like an icon or underline) is provided so the color contrast between plain and link text is supplementary, what Lc is considered sufficient for a user to be reasonably able to discern that two colors are different?

Thanks again!
Colin

**Reply by front-endian** [reply]
I'll add that this question came up from the following issue:

We were designing a dark mode color scheme. The text and link color needed to be Lc 75 against the black background, then we assumed that because we saw Lc 60 listed as the equivalent of 3:1 for readability of text, the contrast ratio between the link text and plain text colors should be Lc 60. We discovered that there were no colors that could give us a Lc of 75 against the background for readability, and an Lc of 60 compared to one another, thus the question of which Lc should be used as a minimum for "users can tell that these two objects are different colors if they are are show on two different parts on the same page".

**Comment by Myndex** [comment]
Hi Colin @justcolin 

I can't give a complete reply right now, but I wanted to mention that luminance contrast is not recommended for information coding, either — in fact as far as information coding, luminance is less functional than hue/chroma. Just FYI, I generally object to all of the 1.4.x SCs as they pertain to contrast, which is why I am doing this work. There is an unfortunate lack of science or empirical data to support most of them, and I have contrary findings as does the existing literature. There seems to have been a move to "just slap 3:1 on everything" — but while 3:1 is the classic readability contrast metric, that does not necessarily apply across the board to all stimuli types, and it is wrong to make such a blanket assumption.

Luminance is what carries fine details, so high luminance contrast is needed for text for readability. But the human vision system is not necessarry good at discriminating luminance _difference_. The FAA limits luminance coding to 3: black grey white.

When we are talking about underlines and bodytext, we have a host of issues. Is readability important? Because underlines interfere with readability. 

But also, when we think of information coding, we're usually meaning _"north bound busses are red, east bound are blue, cross town are orange..."_ Inline text links are no where near that level of information coding. The only real question is how distinct is it from the non-link text? This level of distinction does not require the same level of consideration as that of multi-item information coding.

Here then are some examples. 

![Example of blue in-line link text](https://user-images.githubusercontent.com/42009457/152612345-1a197dc7-43d4-4cfe-89b3-fead4ad6e139.png)

The classic blue link has a great deal of value. In this example, the blue links are Lc75, and the black text is Lc97, for about a Lc20 difference, which by the way... is 3:1 under WCAG 2.

### _Wait WUT!?_

Remember how we discussed that as colors get darker WCAG2 contrast becomes less and less accurate? Well here's how we can work that into our favor. And dealing with things like 1.4.1 or 1.4.11, there are going to be times (if you have to meet WCAG 2) where it is advantageous just to use _that_ math, *because* Bridge PCA is about making **text readable**. The other guidelines that surround APCA that make it flexible had to be tossed out, in order to facilitate backwards compatibility.

Now, let's see how this plays out for color vision:

<img width="870" alt="Correct CVD sim of blue text links" src="https://user-images.githubusercontent.com/42009457/152612527-dae9230a-924c-494f-8497-72ce0bc1818a.png">

All forms of color vision deficiency can both read and distinguish the links from the surrounding text. Trying to do this by just making the text grey, even at Lc 75 vs Lc97, does not work particularly well.

1.4.1 does not say to discard color information, only to augment it.

And you'll notice that the color chosen is #0050c2 - blue with a tiny hint of green. And guess what? Deutan. protan and tritan ALL see green.

### _Wait WUT again?_

Deuteranopia/deuteranomalous, are missing or have problems with the M cones, sometimes called the "green" cones. But wait, if they are missing, how do they see green? 

Well, the reason we call them M for medium wavelength, and not "green" is that both the L cone (red) and the M cone overlap, and they **both have peak response in green**. (greenish/yellow for L and yellowish/green for M).

The L cone overlaps the M cone almost completely, and the L cone extends down into red. So when the M cone is missing, the L cone takes up the slack to the point that deutans have essentially standard luminance contrast sensitivity.

### Summing up

So, we know we need luminance contrast for small, thin things like text, for readability.

But we also need to understand the needs regarding discrimination, object detection, motion... those all have different requirements, and are all processed by different areas of the brain and visual cortex than lexical processing of luminance information.

I hope I didn't go too far tangent, but try #0050c2 for links in text against #000 text. It meets WCAG2, and should more than meet future APCA guidelines in this narrow use case.

Thank you!

**Reply by TaigaYamada** [reply]
Finer levels of information coding with luminance being discouraged is fascinating. I didn't know that at all. 


So how should one approach 5-6 levels of information coding? Vary luminance/chroma/hue at the same time?


In the future, guidance aimed at data visualization will be very helpful, especially because they are increasingly common

**Reply by Myndex** [reply]
Hi @TaigaYamada 
> Finer levels of information coding with luminance being discouraged is fascinating. I didn't know that at all.

We are pretty good a recognizing luminance differences on _adjacent_ colors. But we are terrible at judging differences in when those same colors are in different context.

![An image with two yellow dots, one of them is in shadow and even though they are exactly the same color on the screen, they both look very different due to a different context. Yellow Dot Checker Shadow_DLyon_PD](https://user-images.githubusercontent.com/42009457/152644082-41ed285c-3f68-4342-a1c8-2a216acc2bce.png)

See the two yellow dots, one of them is in shadow and even though they are **_exactly the same color on the screen_**, they both look very different due to the different context.

Both yellow dots are ` #d18600 ` and they both sit on squares that are ` #717171 `.

It should be noted that we also have context sensitive hue and chroma, but we also have a greater capacity for recognizing and distinguishing differences in hue than we do for luminance. Unless of course we are color vision insensitive.


> So how should one approach 5-6 levels of information coding? Vary luminance/chroma/hue at the same time?

When you vary hue and chroma, you will essentially always be varying luminance, unless you are using a color model with orthogonal lightness.

But the general guidance is to:
- Use colors that are distinguishable by color insensitive.
- Use at least one additional marker, be it a call-out, or symbol, or variation in spatial frequency properties.

Eg: let's say you have a line chart, and you want to make one of the lines lighter. Instead of lowering luminance, consider making it dotted, or thinner. Thinner = lower contrast. Thicker = higher contrast.

> In the future, guidance aimed at data visualization will be very helpful, especially because they are increasingly common

DataViz is definitely part of our research. It's just the text issue has been the one that is most forefront because that is the one that is the current critical crisis.

**Reply by TaigaYamada** [reply]
Thank you very much for the clarification!

**Reply by front-endian** [reply]
Thank you, this is all very helpful!

Before marking it as answered, let me confirm that I am understanding the gist of your answer:
- APCA and SAPC are designed to measure readability of text, or the ability to clearly distinguish different shapes/details in graphics
- APCA and SAPC can **not** be used to measure how easily users can tell the difference between two different colored objects
- While WCAG 2 mentions using a 3:1 contrast measurement as a minimum for whether a person can tell the difference between two colors, this is faulty
  - There was no scientific basis for that WCAG recommendation
  - Humans are bad at perceiving contrast differences between two colors, and it is is very context dependent (this is the basis of many optical illusions)
- The same old rule of thumb still applies: always use at least one non-color (shape, texture, icon, label) way of encoding any information, then make sure colors you choose at least look different for people with various color vision deficiencies.

**Reply by Myndex** [reply]
hi Colin @justcolin 

> * _APCA and SAPC are designed to measure readability of text, or the ability to clearly distinguish different shapes/details in graphics_

Yes.
> * _APCA and SAPC can **not** be used to measure how easily users can tell the difference between two different colored objects_

APCA operates on luminance only, it is a derivation from the larger SAPC/SARCAM model. SAPC (full model) can definitely determine hue/chroma/saturation differences. SAPC is not part of anything currently publicly facing, and is not part of anything (at present) being donated to the W3.

> * _While WCAG 2 mentions using a 3:1 contrast measurement as a minimum for whether a person can tell the difference between two colors, this is faulty_
>   * _There was no scientific basis for that WCAG recommendation_
>   * _Humans are bad at perceiving contrast differences between two colors, and it is is very context dependent (this is the basis of many optical illusions)_

Correct, context is key.

> * _The same old rule of thumb still applies: always use at least one non-color (shape, texture, icon, label) way of encoding any information, then make sure colors you choose at least look different for people with various color vision deficiencies._

Yes, this predates computers by at least decades, old maps are often good examples with color+luminance+physical/dimensional differences—classical design rules are classical for a reason 😎… The reasonable exception is when there's interference with a higher level requirement such as readability. And this is where we get into the tender area of in-line text links as I demonstrated above.

**Reply by front-endian** [reply]
Fantastic! Thank you for clarifying all this.

**Comment by clarfonthey** [comment]
Just want to say that I was poking around for some more information on this, and the explanation you gave with the link colours was very helpful. I guess that it's kind of hard when you're stuck in the maths to just remember why colour exists in the first place; you can't really differentiate more than two or three things by contrast, but colour helps distinguish more than that. So, since not everyone sees colour the same way, the key is to design things that either a) don't need to distinguish between too many things, or b) use something other than colour to distinguish, like bolding, underlines, etc.

**Comment by Garz** [comment]
After looking around for a bit after posting, I think this discussion helps bring a bit of clarity to my question. If i followed correctly, the contrast between a button fill color and the background it sits on is not something APCA is trying to define nor should we try and hit those contrast levels for that particular case?

**Reply by Myndex** [reply]
It is something that APCA can and will do, though we haven't fleshed out the guidelines.

**Reply by Garz** [reply]
Is there anything in the current set of guidelines to use for now to guide our decision making? WCAG is not very useful in the scenario of building UI as it relates to interactive states and fills (Not related to text)

Ex: Light Mode Button - Fill: #000 at 14% opacity / Border: #000 at 18% opacity. 

Our goal at the moment is to land between 15 and 30 on APCA contrast levels. Ideally closer to 30 as long as the fills don't effect visual clarity and contrast when text is layered on top.

Something that we are working through is how do we balance the contrast between a UI element fill as it sits on a page, and when text is within that button. Raising the fill too high effects text contrast and the general presentation of an experience.
